### PR TITLE
Add deep copy for httpclient config

### PR DESCRIPTION
**Pull request by jdillon**
copy/clone pattern used to protect runtime state, but was not implemented for httpclient config, this adds deep copy to prevent corruption of configured state.

### CODE REVIEW

**Comment by jdillon** [comment]
http://bamboo.s/browse/NX3-OSSF372-1 :white_check_mark:

File: components/nexus-httpclient/src/main/java/org/sonatype/nexus/httpclient/config/ProxyConfiguration.java [hunk 1]
```diff
@@ -63,6 +65,25 @@ public void setNonProxyHosts(@Nullable final String[] nonProxyHosts) {
...
+    try {
+      ProxyConfiguration copy = (ProxyConfiguration)clone();
+      if (http != null) {
+        copy.http = http.copy();
+      }
+      if (https != null) {
+        copy.https = https.copy();
+      }
+      if (nonProxyHosts != null) {
+        copy.nonProxyHosts = Arrays.copyOf(nonProxyHosts, nonProxyHosts.length);
```
Comment (mcculls) [inline components/nexus-httpclient/src/main/java/org/sonatype/nexus/httpclient/config/ProxyConfiguration.java]: FYI, Java arrays are implicitly cloneable so you could also use `copy.nonProxyHosts = nonProxyHosts.clone();`

(note this is a shallow clone, it does not create copies of individual elements)

**Comment by mcculls** [comment]
+1

File: components/nexus-httpclient/src/main/java/org/sonatype/nexus/httpclient/config/ProxyConfiguration.java (hunk 1 above)
Comment (jdillon) [inline components/nexus-httpclient/src/main/java/org/sonatype/nexus/httpclient/config/ProxyConfiguration.java]: good to know, will update.  This a String[] so nothing deeper needed in this case.

File: components/nexus-httpclient/src/main/java/org/sonatype/nexus/httpclient/config/AuthenticationConfiguration.java [hunk 2]
```diff
@@ -33,4 +35,13 @@ public String getType() {
   }
 
   // TODO: preemptive?
```
Comment (mrprescott) [inline components/nexus-httpclient/src/main/java/org/sonatype/nexus/httpclient/config/AuthenticationConfiguration.java]: Is this TODO still relevant?

Comment (jdillon) [inline components/nexus-httpclient/src/main/java/org/sonatype/nexus/httpclient/config/AuthenticationConfiguration.java]: depends of if we want to add a flag to enable preemptive auth or not.  I assume we did but wasn't sure how that was setup presently in the codebase.

**Comment by mrprescott** [comment]
+1

**Comment by jdillon** [comment]
manually merged

**Pull request closed without merging**